orbit-2-livedata/src/main/java/com/babylon/orbit2/DelegatingLiveData.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2020 Babylon Partners Limited
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.babylon.orbit2
+
+import androidx.annotation.MainThread
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Observer
+import java.io.Closeable
+
+/**
+ This class creates one LiveData per observer in order to defer to the behaviour of the [Stream]
+ when it comes to caching values. This ensures that state and side effect caching is properly
+ resolved while retaining the benefits of using LiveData in terms of main thread callbacks and
+ automatic unsubscription.
+ */
+internal class DelegatingLiveData<T>(private val stream: Stream<T>) : LiveData<T>() {
+    private val closeables = mutableMapOf<Observer<in T>, LiveData<T>>()
+
+    @MainThread
+    override fun observe(owner: LifecycleOwner, observer: Observer<in T>) {
+        // Observe the internal MutableLiveData
+        closeables[observer] = InternalStreamLiveData(stream).also {
+            it.observe(owner, observer)
+        }
+    }
+
+    override fun observeForever(observer: Observer<in T>) {
+        closeables[observer] = InternalStreamLiveData(stream).also {
+            it.observeForever(observer)
+        }
+    }
+
+    override fun getValue(): T? {
+        throw UnsupportedOperationException(
+            "This operation is not supported. Call Container.currentState instead."
```

Comment:
what does this exception mean from a practical point of view?

also there's no test for it

Reply:
I think it's pretty inconsequential, but basically if someone tried to `getValue` it would always return null, even though the state is set (because the actual streaming is handled by a delegate). So this is to discourage people from using this and rather pull it out of the container instead.

Reply:
so you get an update over livedata but cannot get the state value from it? does this not end up feeling a little too non-standard then?

Reply:
it is, but I couldn't find a way to turn off emit-state-on-connect without circumventing caching altogether. The `LiveData` API is standard and nice so I wanted to keep it.

There are two alternatives:

1. Keep caching as normal for state, but turn it off for side effects only
2. Ditch LiveData altogether and build our own API around lifecycles. This is a good alternative as it requires less hacky magic to achieve, all we need is a new kind of `observe` function that reacts to lifecycle changes to auto-unsubscribe.